public/js/utils/source-map.js

```diff
@@ -1,160 +1,206 @@
 }
 
-const sourceMapTask = function(method) {
-  return function() {
-    const args = Array.prototype.slice.call(arguments);
-    return workerTask(sourceMapWorker, { method, args });
-  };
-};
+/**
+ * Sets the source map's sourceRoot to be relative to the source map url.
+ */
+function _setSourceMapRoot(sourceMap, absSourceMapURL, source) {
+  // No need to do this fiddling if we won't be fetching any sources over the
+  // wire.
+  if (sourceMap.hasContentsOfAllSources()) {
+    return;
+  }
+
+  const base = path.dirname(
+    (absSourceMapURL.indexOf("data:") === 0 && source.url) ?
+      source.url :
+      absSourceMapURL
+  );
 
-const getOriginalSourcePosition = sourceMapTask("getOriginalSourcePosition");
-const getGeneratedSourceLocation = sourceMapTask("getGeneratedSourceLocation");
-const createOriginalSources = sourceMapTask("createOriginalSources");
-const getOriginalSourceUrls = sourceMapTask("getOriginalSourceUrls");
-const getOriginalTexts = sourceMapTask("getOriginalTexts");
-const createSourceMap = sourceMapTask("createSourceMap");
-const clearData = sourceMapTask("clearData");
+  if (sourceMap.sourceRoot) {
+    sourceMap.sourceRoot = path.join(base, sourceMap.sourceRoot);
+  } else {
+    sourceMap.sourceRoot = base;
+  }
 
-function _shouldSourceMap(source) {
-  return isEnabled("sourceMaps") && source.sourceMapURL;
+  return sourceMap;
 }
 
-function isMapped(source) {
-  return _shouldSourceMap(source);
+function originalToGeneratedId(originalId) {
+  const match = originalId.match(/(.*)\/originalSource/);
+  return match ? match[1] : null;
 }
 
-function isOriginal(originalSource) {
-  return !!getGeneratedSourceId(originalSource);
+function generatedToOriginalId(generatedId, url) {
+  return generatedId + "/originalSource-" + md5(url);
 }
 
-function isGenerated(source) {
-  return !isOriginal(source);
+function isOriginalId(id) {
+  return id.match(/\/originalSource/);
 }
 
-async function getOriginalSources(state: AppState, source: any) {
-  const originalSourceUrls = await getOriginalSourceUrls(source);
-  return originalSourceUrls.map(url => getSourceByURL(state, url));
+function isGeneratedId(id) {
+  return !isOriginalId(id);
 }
 
-function getGeneratedSource(state: AppState, source: any) {
-  if (isGenerated(source)) {
-    return source;
+function fetchSourceMap(generatedSource) {
+  // If it has already been requested, return the request. An
+  // important behavior here is that if it's in the middle of
+  // requesting it, all subsequent calls will block on the initial
+  // request.
+  const existingRequest = sourceMapRequests.get(generatedSource.id);
+  if (existingRequest) {
+    return existingRequest;
+  } else if (!generatedSource.sourceMapURL || !isEnabled("sourceMaps")) {
+    return Promise.resolve(null);
   }
 
-  const generatedSourceId = getGeneratedSourceId(source);
-  const originalSource = getSource(state, generatedSourceId);
+  async function request() {
```

Comment:
Engines definitely optimize this :) I like how it's outlined and groups together the code instead of having two top-level functions, but I don't feel that strongly about it.